src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -8,393 +8,108 @@
-	} else {
-		dim_input_space = num_features;
-		// does not reinit if dimension is the same to avoid overriding a previous call of set_randomcoefficients(...)
-		init_randomcoefficients();
-	}
+	auto num_features = f->as<DenseFeatures<float64_t>>()->get_num_features();
+	require(num_features > 0, "Input space dimension must be greater than zero");
+	
+	init_basis(num_features);
+	m_fitted = true;
 }
 
 SGVector<float64_t> RandomFourierGaussPreproc::apply_to_feature_vector(SGVector<float64_t> vector)
 {
-	if (!test_rfinited()) {
-		throw ShogunException(
-				"float64_t * CRandomFourierGaussPreproc::apply_to_feature_vector(...): test_rfinited()==false: you need to call before CRandomFourierGaussPreproc::init (Features *f) OR	1. set_dim_feature_space(const int32 dim), 2. set_dim_input_space(const int32 dim), 3. init_randomcoefficients() or set_randomcoefficients(...) \n");
-	}
-
-	float64_t val = std::sqrt(2.0 / cur_dim_feature_space);
-	SGVector<float64_t> res(cur_dim_feature_space);
+	SGVector<float64_t> projection(m_dim_output);
+	SGVector<float64_t> offset_vector(m_dim_output);
 
-	for (int32_t od = 0; od < cur_dim_feature_space; ++od) {
-		SGVector<float64_t> wrapper(randomcoeff_multiplicative+od*cur_dim_input_space, cur_dim_input_space, false);
-		res[od] = val * cos(randomcoeff_additive[od] + linalg::dot(vector, wrapper));
-	}
+	assert_fitted();
+	linalg::add(m_offset, linalg::matrix_prod(m_basis, vector), offset_vector);
+	projection = linalg::scale(linalg::cos(offset_vector), std::sqrt(2.0 / m_dim_output));
 
-	return res;
+	return projection;
 }
 
-SGMatrix<float64_t>
-RandomFourierGaussPreproc::apply_to_matrix(SGMatrix<float64_t> matrix)
+SGMatrix<float64_t> RandomFourierGaussPreproc::apply_to_matrix(SGMatrix<float64_t> matrix)
 {
-	// version for case dim_feature_space < dim_input space with direct transformation on feature matrix ??
-
 	auto num_vectors = matrix.num_cols;
 	auto num_features = matrix.num_rows;
 
-	io::info("get Feature matrix: {}x{}", num_vectors, num_features);
+	assert_fitted();
+	SG_DEBUG("Got Feature matrix: {}x{}", num_vectors, num_features);
 
-	if (num_features != cur_dim_input_space)
-	{
-		throw ShogunException(
-		    "float64_t * "
-		    "CRandomFourierGaussPreproc::apply_to_matrix("
-		    "SGMatrix<float64_t> matrix): matrix.num_rows != "
-		    "cur_dim_input_space is not allowed\n");
-	}
+	SGMatrix<float64_t> projection(m_dim_output, num_vectors);
 
-	SGMatrix<float64_t> res(cur_dim_feature_space, num_vectors);
+	/* Temporary matrix to store offsetted matrix product*/
+	SGMatrix<float64_t> offset(m_dim_output, num_vectors);
 
-	auto val = std::sqrt(2.0 / cur_dim_feature_space);
+	linalg::add_vector(
+	    linalg::matrix_prod(m_basis, matrix), m_offset, offset);
+	projection = linalg::scale(linalg::cos(offset), std::sqrt(2.0 / m_dim_output));
+
+	return projection;	
+}
 
-	for (auto vec : range(num_vectors))
+void RandomFourierGaussPreproc::sample_spectral_density(int32_t dim_input_space)
```

Comment:
this is not what I meant with putting the spectral density into a method, and it doesnt just do that, it does more.
Could you make the method only sample from the kernel, i.e. it takes a matrix, and just fills it with values from the kernel density? (Also doesnt modify class members)
Then the offset code, you can do that above where you call this method atm
The goal of outsourcing this into a sep method is that we then can move on to make this happen for various kernel functions

Reply:
Thanks, I think I get the picture. Shall I create another method for offsetting? It would be ugly to start a loop inside ```init_basis```.

Reply:
no I think you can just do exactly that. No problem.

Reply:
Silly question, just take dim_input_space and return a matrix rather than "matrix as arg" then filling it?
so that we could have
```
m_basis = sample_spectral_density(dim_input_space);
```